src/app.js

```diff
@@ -1,12 +1,131 @@
+import { AsyncList } from './asyncList.js';
+import { CoffeeShop } from './coffeeShop.js';
+
 /**
- * @param {Object} position
- * @param {Number} position.x
- * @param {Number} position.y
- * 
- * @returns {Array<position>}
+ * @typedef {object} Point
+ * @property {number} x
+ * @property {number} y
  */
-export function getNearestShops(position) {
-  // code
-  
-  return [];
+
+/**
+ * @callback getNearestShops
+ * @param {Point} position
+ * @param {number} numberOfShopsToReturn
+ * @returns {Promise<Array<import('./coffeeShop.js').ShopDistance>>}
+ */
+
+/**
+ * @param {fetchShopList} fetchShops
+ * @param {import("./coffeeShop.js").shopDistanceTo} shopDistanceTo
+ * @param {import("./coffeeShop").distanceAsc} distanceAsc
+ * @returns {getNearestShops}
+ */
+export function nearestShopsFactory(fetchShops, shopDistanceTo, distanceAsc) {
+    return (position, numberOfShopsToReturn) => fetchShops()
+        .map(shopDistanceTo(position))
+        .sort(distanceAsc)
+        .take(numberOfShopsToReturn);
+}
+
+/**
+ * @callback fetchShopList
+ * @returns {import('./asyncList.js').AsyncList}
+ */
+
+/**
+ * @param {fetchAPIToken} fetchAPIToken
+ * @param {fetchCoffeeShops} fetchShops
+ * @returns {fetchShopList}
+ */
+export function fetchShopListFactory(fetchAPIToken, fetchShops) {
+    return () => new AsyncList(fetchAPIToken().then(fetchShops));
 }
```

Comment:
The asyncList structure is not needed. I could use `then()` instead of `map` `sort` and `take`. I could make a case for readability however at the cost of complexity. In a bigger project that tradeoff may be good however there are libraries for this no need to reinvent the wheel. I was trying to show off.